[14.0][FIX] stock_picking_filter_lot configurable per operation_type

remi-filament opened the pull request:
I have finally found a solution for #865 

This PR proposes to :
- add boolean use_filter_lots on stock_picking_type
- apply filter on lots only for picking where picking_type has use_filter enabled

remi-filament commented:
Just added updates on README I forgot in initial commit

bealdav commented on stock_picking_filter_lot/models/stock_move_line.py:
```diff
@@ -0,0 +1,38 @@
+                        ("product_id", "=", rec.product_id.id),
+                        ("company_id", "=", rec.company_id.id),
+                    ]
+                )
```
Ok I understand the topic here.
Please you may avoid to have these duplicates lines by adding conditionnally ("location_ids", "child_of", rec.location_id.id) to first 2 tuples and send to json.dumps

gdgellatly commented:
Hi

Are we sure this is the way to go? It seems really complicated way to solve a simple issue. And usability wise it isn't nice. Not to mention merging will break existing installations. Then even when they do get the dependencies, they need to update all existing picking types, surely the default must be true here. 

The problem only presents in a non standard configuration, mostly for receiving existing lots. Personally I feel this change needs to be its own module. But I also feel we are missing a much easier solution. I just can't think of it but it revolves around location_id.usage.

Also computed and related fields are readonly by default. There is no need to set that.

remi-filament commented:
@gdgellatly I have updated code with your comment to get default to true in order not to disturb existing installations and removed redundant readonly and store = False.
I tend to think also that there should be a better way to implement this but I could not get to it (but I am no Python expert so I am most probably missing something).

Fine with me if you prefer this to go in another module I can do that also.
Let me know if you can think of a better/easier way to implement it ?

gdgellatly commented:
I had a think about it. For me I'd prefer this picking type stuff is in another module and we do something like this to avoid the dependency. But done without the dependency, and a default of true on the picking types I wouldn't block it for in the main module.

In pseudocode

```python
class MoveLine(...)
    filter_lots = fields.Boolean(related=<picking_type_field>)

class PickingType(..)
  filter_lots = fields.Boolean(default=True)
```

We can just use the boolean value in the view domain.

Then on the views we just add
`<field name='filter_lots', invisible='1' />`

And for the field domain we just have
`[(product),(company)] + (filter_lots and [(location)] or [])`

remi-filament commented:
Thanks for giving this point some thoughts !
I am wondering whether it is possible to set domain as you propose directly in view ?

Otherwise I guess we still need web_domain_field dependency, no ?

Also, unless I miss something (like you can set domain with variable directly in view), I feel like what you propose is quite similar to what I did (except that you propose to make domain with and / or instead of my if), right ?

gdgellatly commented:
https://github.com/odoo/odoo/blob/97e3115bce518cd9d98fe2454e341077077a79a8/addons/purchase/views/account_move_views.xml#L17

Yes you can define in view for simple boolean test and changes.

The difference is this won't break any installations, either functionally or with new dependencies. It will add a field but such is life. That happens in OCA sometimes.

remi-filament committed:
[FIX] stock_picking_filter_lot made configurable

remi-filament commented:
Great I did not know that, thanks for the tip, it looks much better now.
Only one point that still could be improved : I did not manage to add tuples in domain so I had to repeat the product and company part on both sides of the or.

I tried with a + like in python code but got the following error :  unsupported operand type(s) for +: 'tuple' and 'tuple'

Any idea ?

gdgellatly commented:
@remi-filament you don't need to add tuples, you need to add lists of tuples. e.g. [(tuple),(tuple)] + ([(tuple)] or [])

remi-filament commented:
This is what I tried :

- [('product_id','=', product_id), ('company_id', '=', company_id)] + (picking_type_use_filter_lots and [('location_ids', 'child_of', location_id), ] or [])
- [('product_id','=', product_id), ('company_id', '=', company_id)] + (picking_type_use_filter_lots and [('location_ids', 'child_of', location_id)] or [])

Both give me the same error when trying to set lot number : unsupported operand type(s) for +: 'tuple' and 'tuple'

gdgellatly commented:
@remi-filament idk what issue is. Maybe you have a comma wrong somewhere, maybe ORM interprets it differently and converts list to a tuple before evaluating, as first one works fine. Second one I can't tell difference, but it doesn't matter if you go long way. See below for test.

```python
Python 3.8.10 (default, Jun  2 2021, 10:49:15) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> product_id = 4
>>> company_id = 2
>>> picking_type_use_filter_lots = True
>>> location_id = 5
>>> [('product_id','=', product_id), ('company_id', '=', company_id)] + (picking_type_use_filter_lots and [('location_ids', 'child_of', location_id), ] or [])
[('product_id', '=', 4), ('company_id', '=', 2), ('location_ids', 'child_of', 5)]
>>> picking_type_use_filter_lots = False
>>> [('product_id','=', product_id), ('company_id', '=', company_id)] + (picking_type_use_filter_lots and [('location_ids', 'child_of', location_id), ] or [])
[('product_id', '=', 4), ('company_id', '=', 2)]
```

remi-filament commented:
Thanks for looking into it, using plus on list of tuples is properly handled in Python code but not un XML view...
I guess we can go for it as is, it is already much better than my original idea thanks to your help !

remi-filament commented:
Just rebased on 14.0

rousseldenis reviewed (approved):
Code review

rousseldenis commented:
/ocabot merge major

rousseldenis commented:
@remi-filament Do you plan to forwardport this ?

remi-filament commented:
Hi @rousseldenis I never used v15 so far, so I am not feeling comfortable to forward-port to v15

rousseldenis commented:
> Hi @rousseldenis I never used v15 so far, so I am not feeling comfortable to forward-port to v15

You can basically use oca-port to do it and make a PR. If it works as is, fine !
